File: src/scanners/shebang.ts
import type { Finding } from "../types";
import { makeFinding } from "../util";

/** Rewrite shebangs using node → bun. */
export function scanShebang(path: string, content: string): { findings: Finding[]; content: string } {
  const findings: Finding[] = [];
  if (!content.startsWith("#!")) return { findings, content };
  const firstLine = content.slice(0, content.indexOf("\n"));
  // #!/usr/bin/env node   or   #!/usr/bin/node   or   #!node
  const re = /^#!\s*(?:\/usr\/bin\/env\s+node|\/usr\/bin\/node|node)(?=\s|$)/;
  const match = firstLine.match(re);
  if (!match) return { findings, content };

  const replacement = "#!/usr/bin/env bun";
  findings.push(makeFinding(path, content, 0, "shebang/node",
    `shebang 'node' → 'bun'`, firstLine, replacement, { autofix: true }));
  const rest = content.slice(firstLine.length);
  return { findings, content: replacement + rest };
}

File: crates/n2b-core/src/scanners/shebang.rs
use crate::types::{Finding, MakeFindingOpts};
use crate::util::{line_offsets, make_finding};
use once_cell::sync::Lazy;
use regex::Regex;

static SHEBANG_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^#!\s*(?:/usr/bin/env\s+node|/usr/bin/node|node)(?:\s|$)")
        .expect("invariant: SHEBANG_RE regex literal is valid")
});

pub fn scan_shebang(path: &str, content: &str) -> (Vec<Finding>, String) {
    let mut findings = Vec::new();
    if !content.starts_with("#!") {
        return (findings, content.to_string());
    }
    let first_line_end = content.find('\n').unwrap_or(content.len());
    let first_line = &content[..first_line_end];
    if !SHEBANG_RE.is_match(first_line) {
        return (findings, content.to_string());
    }
    let replacement = "#!/usr/bin/env bun";
    let offsets = line_offsets(content);
    findings.push(make_finding(
        path,
        &offsets,
        0,
        "shebang/node",
        "shebang 'node' → 'bun'",
        first_line.to_string(),
        Some(replacement.to_string()),
        MakeFindingOpts {
            autofix: Some(true),
            ..Default::default()
        },
    ));
    let rest = &content[first_line.len()..];
    (findings, format!("{replacement}{rest}"))
}
